src/PotentialSimulation/ConvergenceAndRefinement.jl

```diff
@@ -88,282 +88,181 @@ function _update_till_convergence!( fssrb::PotentialSimulationSetupRB{T, N1, N2}
+   smaller than `max_diff[i]` (i -> dimension) but that the distances between the ticks stays larger than `minimum_distances[i]`.
+3. Create the new data array for the refined grid and fill it by interpolation of the the initial (coarse) grid.
+"""
+function refine_scalar_potential(p::ScalarPotential{T}, max_diffs::NTuple{3, T}, minimum_distances::NTuple{3, T}; 
+        only2d::Val{only_2d} = Val(size(p.data, 2)==1)) where {T, only_2d}
+    closed_potential = _get_closed_potential(p)
+    new_grid = _create_refined_grid(closed_potential, T.(max_diffs), T.(minimum_distances))
+    new_data = Array{T, 3}(undef, size(new_grid))
+    if only_2d
+        int = interpolate_closed_potential(closed_potential, only2d)
+        for i3 in axes(new_data, 3)
+            x3 = new_grid.axes[3].ticks[i3]
+            for i1 in axes(new_data, 1)
+                x1 = new_grid.axes[1].ticks[i1]
+                new_data[i1, 1, i3] = int(x1, x3)
             end
         end
-    end
-
-    if typeof(int2).parameters[2] == :open
-        if isodd(length(inds_φ))
-            for iφ in 1:size(potential, 2)
-                if !in(iφ, inds_φ) 
-                    if iφ != size(potential, 2)
-                        if axφ[iφ + 1] - axφ[iφ] > minimum_distance_φ
-                            append!(inds_φ, iφ)
-                            break
-                        end
-                    else 
-                        if cyclic - axφ[iφ] > minimum_distance_φ
-                            append!(inds_φ, iφ)
-                            break
-                        end
-                    end
+    else
+        int = interpolate_closed_potential(closed_potential, only2d)
+        for i3 in axes(new_data, 3)
+            x3 = new_grid.axes[3].ticks[i3]
+            for i2 in axes(new_data, 2)
+                x2 = new_grid.axes[2].ticks[i2]
+                for i1 in axes(new_data, 1)
+                    x1 = new_grid.axes[1].ticks[i1]
+                    new_data[i1, i2, i3] = int(x1, x2, x3)
                 end
             end
         end
-        if isodd(length(inds_φ))
-            b::Bool = true
-            diffs = diff(axφ)
-            while b
-                if length(diffs) > 0
-                    idx = findmax(diffs)[2]
-                    if !in(idx, inds_φ)
-                        append!(inds_φ, idx)
-                        b = false
-                    else
-                        deleteat!(diffs, idx)
-                    end
-                else
-                    for i in eachindex(axφ)
-                        if !in(i, inds_φ) push!(inds_φ, i) end
-                    end
-                    b = false
-                end
-            end
-        end    
-        @assert iseven(length(inds_φ)) "Refinement would result in uneven grid in φ."
     end
+    return _convert_closed_potential(typeof(p), ElectricPotential(new_data, new_grid))

```

Comment:
Can this be solved more elegantly by making `ScalarPotential` a type?
Here, it looks like you first save it as `ElectricPotential` to convert it to whatever you need afterwards (?)

Reply:
I added methods for `ScalarPotential` to make this more general.